File: src/vot1/dashboard/static/js/wallet.js
/**
 * VOT1 Dashboard - Wallet Integration
 * 
 * This module provides client-side functionality for connecting to Phantom Wallet,
 * performing ZK proof operations, and interacting with hex.dev.
 */

// Wallet Integration Module
const VOT1Wallet = (() => {
    // Private variables
    let _connected = false;
    let _wallet = null;
    let _provider = null;
    let _address = null;
    let _network = 'mainnet';
    let _config = {
        zkEnabled: true,
        hexEnabled: true,
        autoConnect: false
    };
    
    // Event listeners
    const _eventListeners = {
        connect: [],
        disconnect: [],
        transaction: [],
        message: []
    };
    
    /**
     * Initialize the wallet module
     * @param {Object} config - Configuration options
     * @return {Promise<boolean>} Success status
     */
    async function init(config = {}) {
        // Merge provided config with defaults
        _config = {..._config, ...config};
        
        try {
            // Check if Phantom is available in the window object
            const isPhantomInstalled = window.phantom && window.phantom.solana && window.phantom.solana.isPhantom;
            
            if (!isPhantomInstalled) {
                console.warn("Phantom wallet is not installed");
                return false;
            }
            
            // Store the provider for future use
            _provider = window.phantom.solana;
            
            // Try to auto-connect if enabled
            if (_config.autoConnect) {
                await connect();
            }
            
            // Add event listeners to Phantom
            _setupEventListeners();
            
            console.log("Wallet module initialized successfully");
            return true;
        } catch (error) {
            console.error("Failed to initialize wallet module:", error);
            return false;
        }
    }
    
    /**
     * Set up event listeners for the wallet provider
     * @private
     */
    function _setupEventListeners() {
        if (!_provider) return;
        
        _provider.on('connect', (publicKey) => {
            _address = publicKey.toString();
            _connected = true;
            
            // Notify all connect listeners
            _notifyEventListeners('connect', { address: _address });
        });
        
        _provider.on('disconnect', () => {
            _address = null;
            _connected = false;
            
            // Notify all disconnect listeners
            _notifyEventListeners('disconnect', {});
        });
        
        _provider.on('accountChanged', (publicKey) => {
            if (publicKey) {
                _address = publicKey.toString();
                _notifyEventListeners('connect', { address: _address });
            } else {
                _address = null;
                _connected = false;
                _notifyEventListeners('disconnect', {});
            }
        });
    }
    
    /**
     * Notify all event listeners of a specific event
     * @param {string} eventName - Name of the event
     * @param {Object} data - Event data
     * @private
     */
    function _notifyEventListeners(eventName, data) {
        if (!_eventListeners[eventName]) return;
        
        _eventListeners[eventName].forEach(listener => {
            try {
                listener(data);
            } catch (error) {
                console.error(`Error in ${eventName} event listener:`, error);
            }
        });
    }
    
    /**
     * Connect to the Phantom wallet
     * @return {Promise<Object>} Connection result
     */
    async function connect() {
        try {
            if (!_provider) {
                throw new Error("Wallet provider not initialized");
            }
            
            // Connect to the wallet
            const response = await _provider.connect();
            _wallet = response;
            _address = response.publicKey.toString();
            _connected = true;
            
            // Also notify the server
            await fetch('/api/wallet/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    operation: 'connect',
                    wallet_type: 'phantom',
                    network: _network,
                    address: _address
                })
            });
            
            console.log("Connected to wallet:", _address);
            return {
                success: true,
                address: _address,
                network: _network
            };
        } catch (error) {
            console.error("Failed to connect to wallet:", error);
            _connected = false;
            _address = null;
            
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Disconnect from the wallet
     * @return {Promise<Object>} Disconnection result
     */
    async function disconnect() {
        try {
            if (!_provider) {
                throw new Error("Wallet provider not initialized");
            }
            
            // Disconnect from the wallet
            await _provider.disconnect();
            _connected = false;
            _address = null;
            
            // Also notify the server
            await fetch('/api/wallet/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    operation: 'disconnect'
                })
            });
            
            console.log("Disconnected from wallet");
            return {
                success: true
            };
        } catch (error) {
            console.error("Failed to disconnect from wallet:", error);
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Sign a message using the connected wallet
     * @param {string} message - Message to sign
     * @return {Promise<Object>} Signing result
     */
    async function signMessage(message) {
        try {
            if (!_connected || !_provider) {
                throw new Error("Wallet not connected");
            }
            
            // Encode the message to Uint8Array
            const encodedMessage = new TextEncoder().encode(message);
            
            // Sign the message with the wallet
            const signature = await _provider.signMessage(encodedMessage, 'utf8');
            
            // Notify event listeners
            _notifyEventListeners('message', {
                message,
                signature
            });
            
            return {
                success: true,
                signature,
                message
            };
        } catch (error) {
            console.error("Failed to sign message:", error);
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Sign a transaction using the connected wallet
     * @param {Object} transaction - Transaction to sign
     * @return {Promise<Object>} Signing result
     */
    async function signTransaction(transaction) {
        try {
            if (!_connected || !_provider) {
                throw new Error("Wallet not connected");
            }
            
            // In a real app, this would create a Solana transaction
            // For this example, we're just simulating the response
            
            // Sign the transaction
            // const signedTransaction = await _provider.signTransaction(transaction);
            
            // Notify event listeners
            _notifyEventListeners('transaction', {
                transaction,
                status: 'signed'
            });
            
            return {
                success: true,
                status: 'signed',
                transaction_id: 'simulated_tx_id'
            };
        } catch (error) {
            console.error("Failed to sign transaction:", error);
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Generate a ZK proof
     * @param {string} circuitId - ID of the ZK circuit
     * @param {Object} inputs - Inputs for the proof
     * @return {Promise<Object>} Proof generation result
     */
    async function generateZKProof(circuitId, inputs) {
        try {
            if (!_config.zkEnabled) {
                throw new Error("ZK functionality is not enabled");
            }
            
            const response = await fetch('/api/wallet/zk', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    operation: 'generate',
                    circuit_id: circuitId,
                    inputs
                })
            });
            
            return await response.json();
        } catch (error) {
            console.error("Failed to generate ZK proof:", error);
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Verify a ZK proof
     * @param {string} circuitId - ID of the ZK circuit
     * @param {Object} proof - The proof to verify
     * @param {Array} publicInputs - Public inputs for verification
     * @return {Promise<Object>} Verification result
     */
    async function verifyZKProof(circuitId, proof, publicInputs) {
        try {
            if (!_config.zkEnabled) {
                throw new Error("ZK functionality is not enabled");
            }
            
            const response = await fetch('/api/wallet/zk', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    operation: 'verify',
                    circuit_id: circuitId,
                    proof,
                    public_inputs: publicInputs
                })
            });
            
            return await response.json();
        } catch (error) {
            console.error("Failed to verify ZK proof:", error);
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Perform a hex.dev operation
     * @param {string} operation - Operation type (encode, decode, hash)
     * @param {string} input - Input data
     * @param {Object} options - Additional options
     * @return {Promise<Object>} Operation result
     */
    async function hexOperation(operation, input, options = {}) {
        try {
            if (!_config.hexEnabled) {
                throw new Error("Hex.dev functionality is not enabled");
            }
            
            const response = await fetch('/api/wallet/hex', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    operation,
                    input,
                    options
                })
            });
            
            return await response.json();
        } catch (error) {
            console.error(`Failed to perform hex.dev ${operation}:`, error);
            return {
                success: false,
                error: error.message
            };
        }
    }
    
    /**
     * Add an event listener
     * @param {string} eventName - Name of the event
     * @param {Function} callback - Event callback function
     */
    function addEventListener(eventName, callback) {
        if (!_eventListeners[eventName]) {
            _eventListeners[eventName] = [];
        }
        
        _eventListeners[eventName].push(callback);
    }
    
    /**
     * Remove an event listener
     * @param {string} eventName - Name of the event
     * @param {Function} callback - Event callback function to remove
     */
    function removeEventListener(eventName, callback) {
        if (!_eventListeners[eventName]) return;
        
        _eventListeners[eventName] = _eventListeners[eventName].filter(
            listener => listener !== callback
        );
    }
    
    /**
     * Check if wallet is connected
     * @return {boolean} Connection status
     */
    function isConnected() {
        return _connected;
    }
    
    /**
     * Get current wallet address
     * @return {string|null} Wallet address or null if not connected
     */
    function getAddress() {
        return _address;
    }
    
    // Public API
    return {
        init,
        connect,
        disconnect,
        signMessage,
        signTransaction,
        generateZKProof,
        verifyZKProof,
        hexOperation,
        addEventListener,
        removeEventListener,
        isConnected,
        getAddress
    };
})();

// Export the module for use in other scripts
window.VOT1Wallet = VOT1Wallet; 

File: src/vot1/dashboard/api/endpoints/wallet.py
"""
Wallet API Endpoints

This module provides API endpoints for blockchain wallet integration,
including Phantom Wallet connectivity, ZK (zero-knowledge) functionality,
and hex.dev integration.
"""

import os
import json
import time
import logging
import secrets
from typing import Dict, Any, Optional, List
from flask import Blueprint, request, jsonify, current_app, g
from flask.views import MethodView

# Set up logging
logger = logging.getLogger(__name__)

# Create Wallet Blueprint
wallet_bp = Blueprint('wallet', __name__, url_prefix='/wallet')

class WalletAPI(MethodView):
    """API endpoint for wallet operations"""
    
    def get(self):
        """Get wallet connection status and information"""
        try:
            wallet_address = request.args.get('address')
            
            if wallet_address:
                # Get info for a specific wallet address
                return self._get_wallet_info(wallet_address)
            else:
                # Return connection status
                return jsonify({
                    "connected": False,
                    "wallet_type": None,
                    "network": None
                })
        
        except Exception as e:
            logger.exception(f"Error getting wallet info: {e}")
            return jsonify({"error": str(e)}), 500
    
    def post(self):
        """Connect to wallet or perform wallet operations"""
        try:
            data = request.json
            
            if not data:
                return jsonify({"error": "No data provided"}), 400
            
            operation = data.get('operation')
            
            if not operation:
                return jsonify({"error": "Operation is required"}), 400
            
            if operation == 'connect':
                return self._connect_wallet(data)
            elif operation == 'disconnect':
                return self._disconnect_wallet(data)
            elif operation == 'sign_message':
                return self._sign_message(data)
            elif operation == 'sign_transaction':
                return self._sign_transaction(data)
            elif operation == 'submit_zk_proof':
                return self._submit_zk_proof(data)
            elif operation == 'hex_encode':
                return self._hex_encode(data)
            else:
                return jsonify({"error": f"Invalid operation: {operation}"}), 400
        
        except Exception as e:
            logger.exception(f"Error performing wallet operation: {e}")
            return jsonify({"error": str(e)}), 500
    
    def _connect_wallet(self, data):
        """Handle wallet connection"""
        wallet_type = data.get('wallet_type', 'phantom')
        network = data.get('network', 'mainnet')
        
        if wallet_type == 'phantom':
            # Simulate successful connection to Phantom wallet
            # In a real implementation, this would be handled client-side with JS
            return jsonify({
                "connected": True,
                "wallet_type": wallet_type,
                "network": network,
                "message": "This is a simulated Phantom wallet connection. In a real app, connection is handled client-side with JavaScript."
            })
        else:
            return jsonify({"error": f"Unsupported wallet type: {wallet_type}"}), 400
    
    def _disconnect_wallet(self, data):
        """Handle wallet disconnection"""
        return jsonify({
            "connected": False,
            "message": "Wallet disconnected successfully"
        })
    
    def _get_wallet_info(self, address):
        """Get information for a specific wallet address"""
        # This would typically query blockchain data
        # For now, return placeholder data
        return jsonify({
            "address": address,
            "balance": "0.0",
            "tokens": [],
            "transactions": []
        })
    
    def _sign_message(self, data):
        """Simulate message signing"""
        message = data.get('message')
        
        if not message:
            return jsonify({"error": "Message is required"}), 400
        
        # Generate a fake signature
        signature = secrets.token_hex(64)
        
        return jsonify({
            "success": True,
            "signature": signature,
            "message": message
        })
    
    def _sign_transaction(self, data):
        """Simulate transaction signing"""
        transaction = data.get('transaction')
        
        if not transaction:
            return jsonify({"error": "Transaction data is required"}), 400
        
        # Generate a fake signature and transaction ID
        signature = secrets.token_hex(64)
        tx_id = secrets.token_hex(32)
        
        return jsonify({
            "success": True,
            "signature": signature,
            "transaction_id": tx_id
        })
    
    def _submit_zk_proof(self, data):
        """Handle ZK proof submission"""
        proof = data.get('proof')
        public_inputs = data.get('public_inputs')
        
        if not proof:
            return jsonify({"error": "ZK proof is required"}), 400
        
        # In a real implementation, this would verify the proof
        # For now, return a success response
        return jsonify({
            "success": True,
            "verified": True,
            "proof_id": secrets.token_hex(16)
        })
    
    def _hex_encode(self, data):
        """Handle hex.dev encoding operations"""
        input_data = data.get('input')
        encoding_type = data.get('type', 'hex')
        
        if not input_data:
            return jsonify({"error": "Input data is required"}), 400
        
        if encoding_type == 'hex':
            # Convert string to hex
            try:
                hex_result = input_data.encode('utf-8').hex()
                return jsonify({
                    "success": True,
                    "result": hex_result,
                    "input": input_data,
                    "encoding": encoding_type
                })
            except Exception as e:
                return jsonify({"error": f"Hex encoding failed: {str(e)}"}), 500
        else:
            return jsonify({"error": f"Unsupported encoding type: {encoding_type}"}), 400


class ZKProofsAPI(MethodView):
    """API endpoint for ZK proof operations"""
    
    def get(self):
        """Get ZK proofs information"""
        try:
            return jsonify({
                "available_circuits": [
                    {
                        "id": "identity",
                        "name": "Identity Proof",
                        "description": "Zero-knowledge proof of identity verification"
                    },
                    {
                        "id": "transaction",
                        "name": "Transaction Proof",
                        "description": "Private transaction with amount hiding"
                    },
                    {
                        "id": "voting",
                        "name": "Anonymous Voting",
                        "description": "Anonymous voting with ZK proofs"
                    }
                ]
            })
        
        except Exception as e:
            logger.exception(f"Error getting ZK proofs info: {e}")
            return jsonify({"error": str(e)}), 500
    
    def post(self):
        """Generate or verify ZK proofs"""
        try:
            data = request.json
            
            if not data:
                return jsonify({"error": "No data provided"}), 400
            
            operation = data.get('operation')
            circuit_id = data.get('circuit_id')
            
            if not operation or not circuit_id:
                return jsonify({"error": "Operation and circuit_id are required"}), 400
            
            if operation == 'generate':
                return self._generate_proof(circuit_id, data)
            elif operation == 'verify':
                return self._verify_proof(circuit_id, data)
            else:
                return jsonify({"error": f"Invalid operation: {operation}"}), 400
        
        except Exception as e:
            logger.exception(f"Error performing ZK operation: {e}")
            return jsonify({"error": str(e)}), 500
    
    def _generate_proof(self, circuit_id, data):
        """Generate a ZK proof (simulated)"""
        inputs = data.get('inputs', {})
        
        # Simulate proof generation
        proof = {
            "pi_a": [secrets.token_hex(32), secrets.token_hex(32), "1"],
            "pi_b": [[secrets.token_hex(32), secrets.token_hex(32)], [secrets.token_hex(32), secrets.token_hex(32)], ["1", "0"]],
            "pi_c": [secrets.token_hex(32), secrets.token_hex(32), "1"],
            "protocol": "groth16",
            "curve": "bn128"
        }
        
        public_inputs = [secrets.token_hex(16) for _ in range(3)]
        
        return jsonify({
            "success": True,
            "proof": proof,
            "public_inputs": public_inputs,
            "circuit_id": circuit_id
        })
    
    def _verify_proof(self, circuit_id, data):
        """Verify a ZK proof (simulated)"""
        proof = data.get('proof')
        public_inputs = data.get('public_inputs')
        
        if not proof or not public_inputs:
            return jsonify({"error": "Proof and public_inputs are required"}), 400
        
        # Simulate verification (always succeeds)
        return jsonify({
            "success": True,
            "verified": True,
            "circuit_id": circuit_id
        })


class HexDevAPI(MethodView):
    """API endpoint for hex.dev operations"""
    
    def get(self):
        """Get hex.dev information"""
        try:
            return jsonify({
                "service_status": "active",
                "available_operations": [
                    {
                        "id": "encode",
                        "name": "Hex Encode",
                        "description": "Convert data to hexadecimal representation"
                    },
                    {
                        "id": "decode",
                        "name": "Hex Decode",
                        "description": "Convert hexadecimal back to original data"
                    },
                    {
                        "id": "hash",
                        "name": "Hash Function",
                        "description": "Create cryptographic hash of input data"
                    }
                ]
            })
        
        except Exception as e:
            logger.exception(f"Error getting hex.dev info: {e}")
            return jsonify({"error": str(e)}), 500
    
    def post(self):
        """Perform hex.dev operations"""
        try:
            data = request.json
            
            if not data:
                return jsonify({"error": "No data provided"}), 400
            
            operation = data.get('operation')
            input_data = data.get('input')
            
            if not operation or not input_data:
                return jsonify({"error": "Operation and input are required"}), 400
            
            if operation == 'encode':
                return self._hex_encode(input_data, data.get('options', {}))
            elif operation == 'decode':
                return self._hex_decode(input_data, data.get('options', {}))
            elif operation == 'hash':
                return self._hex_hash(input_data, data.get('algorithm', 'sha256'))
            else:
                return jsonify({"error": f"Invalid operation: {operation}"}), 400
        
        except Exception as e:
            logger.exception(f"Error performing hex.dev operation: {e}")
            return jsonify({"error": str(e)}), 500
    
    def _hex_encode(self, input_data, options):
        """Encode data to hexadecimal"""
        try:
            prefix = options.get('prefix', '')
            uppercase = options.get('uppercase', False)
            
            if isinstance(input_data, str):
                hex_result = input_data.encode('utf-8').hex()
            else:
                return jsonify({"error": "Input must be a string"}), 400
            
            if uppercase:
                hex_result = hex_result.upper()
            
            return jsonify({
                "success": True,
                "result": f"{prefix}{hex_result}",
                "input": input_data,
                "operation": "encode"
            })
        
        except Exception as e:
            return jsonify({"error": f"Hex encoding failed: {str(e)}"}), 500
    
    def _hex_decode(self, input_data, options):
        """Decode hexadecimal to data"""
        try:
            strip_prefix = options.get('strip_prefix', True)
            
            if not isinstance(input_data, str):
                return jsonify({"error": "Input must be a string"}), 400
            
            # Remove '0x' prefix if present and requested
            if strip_prefix and input_data.startswith('0x'):
                input_data = input_data[2:]
            
            try:
                decoded = bytes.fromhex(input_data).decode('utf-8')
                return jsonify({
                    "success": True,
                    "result": decoded,
                    "input": input_data,
                    "operation": "decode"
                })
            except Exception:
                # If UTF-8 decoding fails, return the raw bytes as a list of integers
                decoded_bytes = list(bytes.fromhex(input_data))
                return jsonify({
                    "success": True,
                    "result": decoded_bytes,
                    "is_binary": True,
                    "input": input_data,
                    "operation": "decode"
                })
        
        except Exception as e:
            return jsonify({"error": f"Hex decoding failed: {str(e)}"}), 500
    
    def _hex_hash(self, input_data, algorithm):
        """Hash data with specified algorithm"""
        try:
            import hashlib
            
            if not hasattr(hashlib, algorithm):
                return jsonify({"error": f"Unsupported hash algorithm: {algorithm}"}), 400
            
            if isinstance(input_data, str):
                input_bytes = input_data.encode('utf-8')
            else:
                return jsonify({"error": "Input must be a string"}), 400
            
            # Get the hash algorithm
            hash_func = getattr(hashlib, algorithm)
            hash_result = hash_func(input_bytes).hexdigest()
            
            return jsonify({
                "success": True,
                "result": hash_result,
                "input": input_data,
                "algorithm": algorithm,
                "operation": "hash"
            })
        
        except Exception as e:
            return jsonify({"error": f"Hashing failed: {str(e)}"}), 500


# Register the endpoints
wallet_view = WalletAPI.as_view('wallet_api')
zk_proofs_view = ZKProofsAPI.as_view('zk_proofs_api')
hex_dev_view = HexDevAPI.as_view('hex_dev_api')

wallet_bp.add_url_rule('/', view_func=wallet_view, methods=['GET', 'POST'])
wallet_bp.add_url_rule('/zk', view_func=zk_proofs_view, methods=['GET', 'POST'])
wallet_bp.add_url_rule('/hex', view_func=hex_dev_view, methods=['GET', 'POST']) 